storage/badger/events.go

```diff
@@ -6,126 +6,139 @@ import (
 	"github.com/dgraph-io/badger/v2"
 
 	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
 	"github.com/onflow/flow-go/storage"
 	"github.com/onflow/flow-go/storage/badger/operation"
 )
 
 type Events struct {
-	db *badger.DB
+	db    *badger.DB
+	cache *Cache
 }
 
-func NewEvents(db *badger.DB) *Events {
+func NewEvents(collector module.CacheMetrics, db *badger.DB) *Events {
+	retrieve := func(key interface{}) func(tx *badger.Txn) (interface{}, error) {
+		blockID := key.(flow.Identifier)
+		var events []flow.Event
+		return func(tx *badger.Txn) (interface{}, error) {
+			err := db.View(operation.LookupEventsByBlockID(blockID, &events))
+			return events, handleError(err, flow.Event{})
+		}
+	}
+
 	return &Events{
 		db: db,
+		cache: newCache(collector,
+			withStore(noopStore),
+			withRetrieve(retrieve)),
 	}
 }
 
-// Store will store events for the given block ID
-func (e *Events) Store(blockID flow.Identifier, events []flow.Event) error {
-	return operation.RetryOnConflict(e.db.Update, func(btx *badger.Txn) error {
-		for _, event := range events {
-			err := operation.SkipDuplicates(operation.InsertEvent(blockID, event))(btx)
-			if err != nil {
-				return fmt.Errorf("could not insert event: %w", err)
-			}
-		}
-		return nil
-	})
-}
-
 func (e *Events) BatchStore(blockID flow.Identifier, events []flow.Event, batch storage.BatchStorage) error {
-	if writeBatch, ok := batch.(*badger.WriteBatch); ok {
-		for _, event := range events {
-			err := operation.BatchInsertEvent(blockID, event)(writeBatch)
-			if err != nil {
-				return fmt.Errorf("cannot batch insert event: %w", err)
-			}
+	writeBatch := batch.GetWriter()
+	for _, event := range events {
+		err := operation.BatchInsertEvent(blockID, event)(writeBatch)
+		if err != nil {
+			return fmt.Errorf("cannot batch insert event: %w", err)
 		}
-		return nil
 	}
-	return fmt.Errorf("unsupported BatchStore type %T", batch)
+
+	callback := func() {
+		e.cache.Put(blockID, events)
+	}
+	batch.OnSucceed(callback)
+	return nil
 }
 
 // ByBlockID returns the events for the given block ID
 func (e *Events) ByBlockID(blockID flow.Identifier) ([]flow.Event, error) {
-
-	var events []flow.Event
-	err := e.db.View(operation.LookupEventsByBlockID(blockID, &events))
+	tx := e.db.NewTransaction(false)
```

Comment:
We need a Tx even for read?

Reply:
yes, every read has to be wrapped in a transaction, and `db.View is actually creating a tx under the hook as well.

But your comment did draw me attention that we should reuse the tx we created, rather than using db.View which creates a new transaction and throw the one we created away. And this seems to be misused a few places, just fixed them all